### ISSUE TITLE
Failed to start gauge API: Timed out connecting to java (1.0.8)

### ISSUE THREAD

**Issue by haroon-sheikh** [getgauge/gauge#1672]
### Describe the bug
Getting loads of these error when running on the latest version of gauge and gauge-java

### What command(s) did you run when you found the bug?

For e.g. 
```
mvn clean test
```

### Output, stack trace or logs related to the bug 

```sh
Failed to start gauge API: Timed out connecting to java
```

#### Gauge (Output of `gauge -v`)

```sh
Gauge version: 1.1.1
Plugins
-------
html-report (4.0.11)
java (0.7.8)
reportportal (1.6.0)
reportserver (0.2.0)
screenshot (0.0.1)
xml-report (0.2.3)
```

#### Operating System information
OSx

**Comment by sriv** [comment]
what's the gauge-java version in `pom.xml`?  Please ensure that it's `0.7.8`.

Do you see these errors in a new gauge project as well?

**Comment by haroon-sheikh** [comment]
@sriv Yeah gauge-java version is set to 0.7.8 and I cannot reproduce on a newer project.

**Comment by sriv** [comment]
I have [`gauge-tests`](https://github.com/getgauge/gauge-tests) which use gauge, gauge-java, html-report. I ran it against the docker image you mention, using both `1.1.1` and `1.0.8` tags:

`sudo docker run --rm -v `pwd`:/gauge-tests -w="/gauge-tests" --entrypoint="./gradlew" sitture/docker-gauge-java:1.1.1 clean javaFT`

`sudo docker run --rm -v `pwd`:/gauge-tests -w="/gauge-tests" --entrypoint="./gradlew" sitture/docker-gauge-java:1.0.8 clean javaFT`

both of the commands pass for me despite multiple runs (I think I would have made about 6 runs shuffling between versions).

The only difference is that this is a gradle project, (I upgraded my local copy to gradle 6.4) as against yours which is a maven project.

**Comment by cyberjaime45** [comment]
I'm having the same issue.
Here's the gauge logs:
https://paste.ofcode.org/uCFDVryKiMSDHUShBThjJG
Project: Java Maven project:
TOTAL SPECS
97
TOTAL SCENARIO
320

I'm starting to think,  Gauge can't handle big projects

**Comment by haroon-sheikh** [comment]
I agree I think smaller projects seem to work fine. And it seems to be an issue with a project of 200+ specs. @sriv Are you able to try running the docker image `docker run --rm -it sitture/docker-gauge-java:1.1.1` with a large project with `-DinParallel=true`.

**Comment by cyberjaime45** [comment]
On my side, if I downgrade to Gauge v 1.0.8 and Java 0.7.4. Works perfectly fine. I'm able to run in parallel up to 20 nodes

If i use Gauge v 1.0.8 and Java 0.7.5  the ability to run in parallel goes down to 15
If i use Gauge v 1.0.8 and Java 0.7.6, 0.7.7  the ability to run in parallel goes down to 10

I'm falling behind, and I'm afraid I won't be able to use  the latest version.

**Comment by sriv** [comment]
Sorry to hear this @cyberjaime45 @haroon-sheikh 

@cyberjaime45  - in your case I remember our earlier discussion where we had figured that `enable_multithreading` helped run more number of streams. Are you still using multithreading?

However it looks like in this case, the runner isn't coming up at all. 

Can you try this please:

- `cd <project location>`
- `GAUGE_CUSTOM_CLASSPATH=`mvn test-compile gauge:classpath` ~/.gauge/plugins/java/0.7.8/launcher.sh start`

Would you be able to share the output of the above? 

@haroon-sheikh - let me try and get a project with more than 100 specs, the public ones that I know require setting up the application under test which I am not familiar with.

In the meanwhile, could you also try the steps above?

**Comment by cyberjaime45** [comment]
Here's the output:

`WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by com.google.inject.internal.cglib.core.$ReflectUtils$1 (file:/usr/share/maven/lib/guice.jar) to method java.lang.ClassLoader.defineClass(java.lang.String,byte[],int,int,java.security.ProtectionDomain)
WARNING: Please consider reporting this to the maintainers of com.google.inject.internal.cglib.core.$ReflectUtils$1
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future release
WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by com.google.inject.internal.cglib.core.$ReflectUtils$1 (file:/usr/share/maven/lib/guice.jar) to method java.lang.ClassLoader.defineClass(java.lang.String,byte[],int,int,java.security.ProtectionDomain)
WARNING: Please consider reporting this to the maintainers of com.google.inject.internal.cglib.core.$ReflectUtils$1
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future release
WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by com.google.inject.internal.cglib.core.$ReflectUtils$1 (file:/usr/share/maven/lib/guice.jar) to method java.lang.ClassLoader.defineClass(java.lang.String,byte[],int,int,java.security.ProtectionDomain)
WARNING: Please consider reporting this to the maintainers of com.google.inject.internal.cglib.core.$ReflectUtils$1
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future release`

**Comment by sriv** [comment]
I think that is the reason, it should start logging something like:

```
{"logLevel":"debug","message":"Scanning packages for steps"}
{"logLevel":"debug","message":"Loading Open the Gauge homepagevia reflected sources."}
{"logLevel":"debug","message":"Loading Go to Gauge Get Started Pagevia reflected sources."}
{"logLevel":"debug","message":"Loading Ensure installation instructions are availablevia reflected sources."}
{"logLevel":"debug","message":"Scanning packages for hooks"}
{"logLevel":"debug","message":"Scanned steps: "}
{"logLevel":"debug","message":"\tOpen the Gauge homepage : null"}
{"logLevel":"debug","message":"\tGo to Gauge Get Started Page : null"}
{"logLevel":"debug","message":"\tEnsure installation instructions are available : null"}
{"logLevel":"info","message":"Listening on port:39571"}
```

I'll dig deeper in sometime.

**Comment by cyberjaime45** [comment]
I updated my previous comment...

**Comment by sriv** [comment]
ok - one more test:

try commenting this line: https://github.com/getgauge/gauge-java/blob/master/bin/launcher.sh#L186

in `~/.gauge/plugins/java/0.7.8/launcher.sh` and check again?

**Comment by cyberjaime45** [comment]
Here's the output:

`cd ~/workspace/QA_automation_test`
GAUGE_CUSTOM_CLASSPATH=`mvn -q test-compile gauge:classpath` ~/.gauge/plugins/java/0.7.8/launcher.sh start`
WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by com.google.inject.internal.cglib.core.$ReflectUtils$1 (file:/usr/share/maven/lib/guice.jar) to method java.lang.ClassLoader.defineClass(java.lang.String,byte[],int,int,java.security.ProtectionDomain)
WARNING: Please consider reporting this to the maintainers of com.google.inject.internal.cglib.core.$ReflectUtils$1
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future release`

**Comment by sriv** [comment]
Ok, thank you for your patience, i'm not at my computer at the moment. 

I'll take a deeper look when i get back. However, in your case this doesnt look like an issue with the size of the project

**Comment by cyberjaime45** [comment]
Thanks. In the mean time I'll be using Gauge 1.0.8. This seems to work fine for me.

**Comment by haroon-sheikh** [comment]
Hi @sriv, as suggested I've commented out getgauge/gauge-java:bin/launcher.sh@master#L186.. And I see the following output:

```sh
Skipping L186
SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
{"logLevel":"info","message":"Listening on port:54620"}
```

I did notice a issue with running `mvn -q test-compile gauge:classpath`, because I have checkstyle plugin configured to log errors to console. The `gauge:classpath` task also returns list of error above the classpath output.

E.g. `mvn -q test-compile gauge:classpath`

```sh
[ERROR] src/test/resources/a/a.properties:[0] (misc) Translation: Key 'sample' missing.
[ERROR] src/test/resources/b/b.properties:[0] (misc) Translation: Key 'sample' missing.
...${classpath}...
```

and when running `GAUGE_CUSTOM_CLASSPATH=mvn -q test-compile gauge:classpath ~/.gauge/plugins/java/0.7.8/launcher.sh start` throws the following error

```
Skipping L186
Error: Could not find or load main class [ERROR]
Caused by: java.lang.ClassNotFoundException: [ERROR]
```

**Comment by cyberjaime45** [comment]
`[ERROR] Unknown lifecycle phase "/home/qauser/.gauge/plugins/java/0.7.8/launcher.sh". You must specify a valid lifecycle phase or a goal in the format <plugin-prefix>:<goal> or <plugin-group-id>:<plugin-artifact-id>[:<plugin-version>]:<goal>. Available lifecycle phases are: validate, initialize, generate-sources, process-sources, generate-resources, process-resources, compile, process-classes, generate-test-sources, process-test-sources, generate-test-resources, process-test-resources, test-compile, process-test-classes, test, prepare-package, package, pre-integration-test, integration-test, post-integration-test, verify, install, deploy, pre-clean, clean, post-clean, pre-site, site, post-site, site-deploy. -> [Help 1]
org.apache.maven.lifecycle.LifecyclePhaseNotFoundException: Unknown lifecycle phase "/home/qauser/.gauge/plugins/java/0.7.8/launcher.sh". You must specify a valid lifecycle phase or a goal in the format <plugin-prefix>:<goal> or <plugin-group-id>:<plugin-artifact-id>[:<plugin-version>]:<goal>. Available lifecycle phases are: validate, initialize, generate-sources, process-sources, generate-resources, process-resources, compile, process-classes, generate-test-sources, process-test-sources, generate-test-resources, process-test-resources, test-compile, process-test-classes, test, prepare-package, package, pre-integration-test, integration-test, post-integration-test, verify, install, deploy, pre-clean, clean, post-clean, pre-site, site, post-site, site-deploy.
    at org.apache.maven.lifecycle.internal.DefaultLifecycleExecutionPlanCalculator.calculateLifecycleMappings (DefaultLifecycleExecutionPlanCalculator.java:249)
    at org.apache.maven.lifecycle.internal.DefaultLifecycleExecutionPlanCalculator.calculateMojoExecutions (DefaultLifecycleExecutionPlanCalculator.java:218)
    at org.apache.maven.lifecycle.internal.DefaultLifecycleExecutionPlanCalculator.calculateExecutionPlan (DefaultLifecycleExecutionPlanCalculator.java:127)
    at org.apache.maven.lifecycle.internal.DefaultLifecycleExecutionPlanCalculator.calculateExecutionPlan (DefaultLifecycleExecutionPlanCalculator.java:145)
    at org.apache.maven.lifecycle.internal.builder.BuilderCommon.resolveBuildPlan (BuilderCommon.java:97)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:111)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:51)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:309)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:194)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:107)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:993)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:345)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:191)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:566)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:289)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:229)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:415)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:356)
[ERROR]
[ERROR]
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/LifecyclePhaseNotFoundException`

**Comment by sriv** [comment]
`mvn gauge:classpath` is something that was introduced recently. So definitely one of the things to suspect.

Could you try replacing the `launcher.sh` with the launcher from `0.7.4`? i.e. 

```
cd ~/.gauge/plugins/java/0.7.8/
mv launcher.sh launcher.sh.bak 
curl -O https://raw.githubusercontent.com/getgauge/gauge-java/v0.7.4/bin/launcher.sh
```

and then try running your specs?

**Comment by cyberjaime45** [comment]
With: 0.74 seems to work, or I get an expected output.

Mchine:
Gauge 1.1.1
Java 0.7.4
Pom file:
0.7.4

``GAUGE_CUSTOM_CLASSPATH=`mvn -q test-compile gauge:classpath` ~/.gauge/plugins/java/0.7.4/launcher.sh start`
/home//workspace/QA_automation_test/src/main/java/services/FlightBoardService.java:3: error: package com.thoughtworks.gauge does not exist
import com.thoughtworks.gauge.Gauge;
                             ^
/home//workspace/QA_automation_test/src/main/java/membersSite/entities/flightboard/FlightCard.java:3: error: package lombok does not exist
import lombok.Getter;
             ^
/home//workspace/QA_automation_test/src/main/java/membersSite/entities/flightboard/FlightCard.java:5: error: package org.openqa.selenium does not exist
import org.openqa.selenium.By;
                          ^
/home//workspace/QA_automation_test/src/main/java/membersSite/entities/flightboard/FlightCard.java:6: error: package org.openqa.selenium does not exist
import org.openqa.selenium.WebElement;
                          ^
/home//workspace/QA_automation_test/src/main/java/membersSite/entities/flightboard/FlightCard.java:7: error: package org.openqa.selenium.support.pagefactory does not exist
import org.openqa.selenium.support.pagefactory.ByChained;
                                              ^
/home//workspace/QA_automation_test/src/main/java/membersSite/entities/flightboard/FlightCard.java:13: error: package junit.framework does not exist
import static junit.framework.TestCase.fail;
                             ^
/home//workspace/QA_automation_test/src/main/java/membersSite/entities/flightboard/FlightCard.java:13: error: static import only from classes and interfaces
import static junit.framework.TestCase.fail;
^
/home//workspace/QA_automation_test/src/main/java/membersSite/entities/flightboard/FlightCard.java:15: error: cannot find symbol
@Getter
 ^
  symbol: class Getter
/home//workspace/QA_automation_test/src/main/java/membersSite/entities/flightboard/FlightCard.java:18: error: cannot find symbol
    private WebElement element;`

**Comment by sriv** [comment]
@cyberjaime45 - I was curious if you used just the launcher from `0.7.4`, but use gauge-java 0.7.8.

**Comment by cyberjaime45** [comment]
I uninstalled all Gauge java first then I installed Java 0.7.4
gauge install java  --version 0.7.4 

Then I ran the command as it appears above

**Comment by sriv** [comment]
ok, so we know gauge `1.1.1` and gauge-java `0.7.8` has issue, but gauge `1.1.1` and gauge-java `0.7.4` works?

Can you try:

```
gauge install java #this should install `0.7.8`
cd ~/.gauge/plugins/java/0.7.8/
mv launcher.sh launcher.sh.bak 
curl -O https://raw.githubusercontent.com/getgauge/gauge-java/v0.7.4/bin/launcher.sh
```

I am interested in knowing if the issue got introduced due to changes in the launcher.

**Comment by haroon-sheikh** [comment]
The issue started happening on 0.7.7 onwards. It’s the changes around classpath in the launcher script.

**Comment by cyberjaime45** [comment]
@sriv  I installed Java 0.7.8 but the manually replaced the launcher by calling 
curl -O https://raw.githubusercontent.com/getgauge/gauge-java/v0.7.4/bin/launcher.sh

I get the same output as earlier when I had installed java 0.7.4
`/home/qauser/workspace/QA_automation_test/src/main/java/services/FlightBoardService.java:3: error: package com.thoughtworks.gauge does not exist
import com.thoughtworks.gauge.Gauge;
                             ^
/home/qauser/workspace/QA_automation_test/src/main/java/membersSite/entities/flightboard/FlightCard.java:3: error: package lombok does not exist
import lombok.Getter;
             ^
/home/qauser/workspace/QA_automation_test/src/main/java/membersSite/entities/flightboard/FlightCard.java:5: error: package org.openqa.selenium does not exist
import org.openqa.selenium.By;`

**Comment by sriv** [comment]
thanks @haroon-sheikh @cyberjaime45 , that also confirms my suspicion. And no wonder I couldn't see the issue in any of our projects, classpath can be very specific.

And I did not have checkstyle enabled like you did @haroon-sheikh - so that was messing up the output.

It's late here, but I'll take this up next week, now that we know the problem is in the launcher, I'll see what's the best way to fix it. Will keep you posted.
